buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java

```diff
@@ -143,6 +153,55 @@ public void plugin(File plugin) {
         plugin(plugin.toURI());
     }
 
+    public void keystore(String key, String value) {
+        addSupplier("Keystore", keystoreSettings, key, value);
+    }
+
+    public void keystore(String key, Supplier<Object> valueSupplier) {
+        addSupplier("Keystore", keystoreSettings, key, valueSupplier);
+    }
+
+    public void setting(String key, String value) {
+        addSupplier("Settings", settings, key, value);
+    }
+
+    public void setting(String key, Supplier<Object> valueSupplier) {
+        addSupplier("Setting", settings, key, valueSupplier);
+    }
+
+    public void systemProperty(String key, String value) {
+        addSupplier("Java System property", systemProperties, key, value);
+    }
+
+    public void systemProperty(String key, Supplier<Object> valueSupplier) {
+        addSupplier("Java System property", systemProperties, key, valueSupplier);
+    }
+
+    public void environment(String key, String value) {
+        addSupplier("Environment variable", environment, key, value);
+    }
+
+    public void environment(String key, Supplier<Object> valueSupplier) {
+        addSupplier("Environment variable", environment, key, valueSupplier);
+    }
+
+    private void addSupplier(String name, Map<String, Supplier<Object>> collector, String key, Supplier<Object> valueSupplier) {
+        requireNonNull(key, name + " key was null when configuring test cluster `" + this + "`");
+        requireNonNull(valueSupplier, name + " value supplier was null when configuring test cluster `" + this + "`");
+        collector.put(key, valueSupplier);
+    }
+
+    private void addSupplier(String name, Map<String, Supplier<Object>> collector, String key, String actualValue) {
+        requireNonNull(actualValue, name + " value was null when configuring test cluster `" + this + "`");
+        addSupplier(name, collector, key, () -> actualValue);
+    }
+
+    private void checkSuppliers(String name, Map<String, Supplier<Object>> collector) {
```

Comment:
My preference is generally to do this stuff in a task graph `whenReady()` callback as it catches misconfiguration earlier than at actual task execution. Perhaps we can expose a single `void verifyConfig()` method here which the `TestClustersPlugin` calls if a test task is included in the task graph? Also means we won't have to do this different times in 4 different places in this class.

Reply:
that's a great idea, and we already have this in place in `freeze`

Reply:
I had to move the checks back from `freeze` . Running ti when the task graph is ready is not good enough as some of the values will not be available until some task that the current one depends on runs. The value was because the s3 fixture port file not being available when calling `s3Fixture.addressAndPort` too early

Reply:
Ah, bummer. Makes sense though. Is it worth at least consolidating this in a single place rather than doing 4 distinctly separate configuration checks?

Reply:
I'm not convinced, if the check has to be delayed, might was well do it right before the value is needed